### PR TITLE
feat: make RegionScanner aware of PartitionRange

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

This is the first part of a parallel table scan with configurable ordering optimization.

Things left to do in the follow-up PRs:
- [x] store total rows, number of row groups in manifest https://github.com/GreptimeTeam/greptimedb/pull/4183
- [ ] refactor related methods, separate the logic of fetching metadata and build parts
- [ ] dispatch different scanner scan policy based on whether the metadata is properly used.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/mito2/src/read/seq_scan.rs [hunk 1]
```diff
@@ -254,7 +416,13 @@ impl RegionScanner for SeqScan {
     }
 
     fn scan_partition(&self, partition: usize) -> Result<SendableRecordBatchStream, BoxedError> {
-        self.scan_partition_opt(Some(partition))
+        // self.scan_partition_impl(partition)
```
Comment (WenyXu) [inline src/mito2/src/read/seq_scan.rs:419]: Better to leave a TODO comment

File: src/mito2/src/read/seq_scan.rs [hunk 2]
```diff
@@ -187,52 +220,159 @@ impl SeqScan {
...
-            while let Some(batch) = reader
-                .next_batch()
+
+            for id in (0..parts_len).skip(partition).step_by(num_partitions) {
+                let maybe_reader = Self::build_merge_reader(
+                    &stream_ctx,
+                    id,
+                    semaphore.clone(),
+                    &mut metrics,
+                )
```
Comment (evenyag) [inline src/mito2/src/read/seq_scan.rs:339]: We need to use the merge reader to merge all `ScanPart` under the same `partition`.

File: src/common/recordbatch/src/util.rs [hunk 3]
```diff
@@ -29,6 +39,91 @@ pub async fn collect_batches(stream: SendableRecordBatchStream) -> Result<Record
     RecordBatches::try_new(schema, batches)
 }
 
+/// A stream that aggregates multiple streams into a single stream.
+pub struct AggregatedRecordBatchStream {
```
[the remark this answers is not in the document]
Comment (waynexia) [reply src/common/recordbatch/src/util.rs:43]: I tried chain first. But our `SendableRecordBatchStream` requires lots of extra methods over `Stream`. You still need a new type to implement them as both `Chain` and `RecordBatchStream` are not from our side.

[the remark this answers is not in the document]
Comment (evenyag) [reply src/common/recordbatch/src/util.rs:43]: What about renaming the stream to `ChainedRecordBatchStream`? `Aggregated` sounds like we also doing some aggregation.

[the remark this answers is not in the document]
Comment (waynexia) [reply src/common/recordbatch/src/util.rs:43]: Okay :\ GPT gives me that name and I thought it looks good

**Review by evenyag — APPROVED** [review]

**Pull request merged**